File: packages/core/src/programmers/llm/LlmApplicationProgrammer.ts
import {
  IJsonSchemaApplication,
  IJsonSchemaTransformError,
  ILlmApplication,
  ILlmFunction,
  ILlmSchema,
  IResult,
  IValidation,
  OpenApi,
} from "@typia/interface";
import { LlmSchemaConverter } from "@typia/utils";
import ts from "typescript";

import { IProgrammerProps } from "../../context/IProgrammerProps";
import { ITypiaContext } from "../../context/ITypiaContext";
import { LiteralFactory } from "../../factories/LiteralFactory";
import { MetadataFactory } from "../../factories/MetadataFactory";
import { TypeFactory } from "../../factories/TypeFactory";
import { MetadataFunction } from "../../schemas/metadata/MetadataFunction";
import { MetadataObjectType } from "../../schemas/metadata/MetadataObjectType";
import { MetadataParameter } from "../../schemas/metadata/MetadataParameter";
import { MetadataSchema } from "../../schemas/metadata/MetadataSchema";
import { ValidateProgrammer } from "../ValidateProgrammer";
import { JsonApplicationProgrammer } from "../json/JsonApplicationProgrammer";
import { LlmSchemaProgrammer } from "./LlmSchemaProgrammer";

/**
 * Generates LLM function calling application from TypeScript class/interface.
 *
 * Converts TypeScript types to {@link ILlmApplication} with function schemas
 * compatible with LLM function calling. Validates type constraints (single
 * object parameter, no dynamic keys) and generates runtime validators.
 *
 * @author Jeongho Nam - https://github.com/samchon
 */
export namespace LlmApplicationProgrammer {
  export interface IProps extends IProgrammerProps {
    config?: Partial<
      ILlmSchema.IConfig & {
        equals: boolean;
      }
    >;
  }

  export interface IWriteProps {
    context: ITypiaContext;
    modulo: ts.LeftHandSideExpression;
    metadata: MetadataSchema;
    config?: Partial<
      ILlmSchema.IConfig & {
        equals: boolean;
      }
    >;
    name?: string;
  }

  export const write = (props: IWriteProps): ts.CallExpression => {
    const application: ILlmApplication.__IPrimitive = writeApplication({
      context: props.context,
      modulo: props.modulo,
      config: props.config,
      metadata: props.metadata,
      name: props.name,
    });

    const typeNode: ts.ImportTypeNode = props.context.importer.type({
      file: "typia",
      name: "ILlmApplication.__IPrimitive",
      arguments: props.name
        ? [ts.factory.createTypeReferenceNode(props.name)]
        : undefined,
    });

    return ts.factory.createCallExpression(
      props.context.importer.internal("llmApplicationFinalize"),
      props.name ? [ts.factory.createTypeReferenceNode(props.name)] : undefined,
      [
        ts.factory.createAsExpression(
          ts.factory.createSatisfiesExpression(
            LiteralFactory.write(application),
            typeNode,
          ),
          typeNode,
        ),
        ts.factory.createObjectLiteralExpression(
          [
            ts.factory.createPropertyAssignment(
              "equals",
              props.config?.equals === true
                ? ts.factory.createTrue()
                : ts.factory.createFalse(),
            ),
          ],
          false,
        ),
      ],
    );
  };

  export const validate = (props: {
    config?: Partial<ILlmSchema.IConfig>;
    metadata: MetadataSchema;
    explore: MetadataFactory.IExplore;
  }): string[] => {
    // the class
    if (props.explore.top === false)
      if (
        props.explore.object === props.metadata?.objects[0]?.type &&
        typeof props.explore.property === "string" &&
        props.metadata.size() === 1 &&
        props.metadata.nullable === false &&
        props.metadata.isRequired() === true &&
        props.metadata.functions.length === 1
      )
        return validateFunction(
          props.explore.property,
          props.metadata.functions[0]!,
        );
      else return LlmSchemaProgrammer.validate(props);

    // top-level type must be a single non-nullable required object
    const output: string[] = [];
    const isValidType: boolean =
      props.metadata.size() === 1 &&
      props.metadata.objects.length === 1 &&
      props.metadata.isRequired() === true &&
      props.metadata.nullable === false;
    if (isValidType === false)
      output.push(
        "LLM application's generic argument must be a class/interface type.",
      );

    // validate each property of the class/interface
    const object: MetadataObjectType | undefined =
      props.metadata.objects[0]?.type;
    if (object !== undefined) {
      if (object.properties.some((p) => p.key.isSoleLiteral() === false))
        output.push(
          "LLM application does not allow dynamic keys on class/interface type.",
        );
      let least: boolean = false; // tracks whether at least one function exists
      for (const p of object.properties) {
        const rawName: string = p.key.getSoleLiteral()!;
        const name: string = JSON.stringify(rawName);
        const value: MetadataSchema = p.value;
        if (value.functions.length) {
          least ||= true;
          if (isValidType === false) {
            if (value.functions.length !== 1 || value.size() !== 1)
              output.push(
                `LLM application's function (${name}) type does not allow union type.`,
              );
            if (value.isRequired() === false)
              output.push(
                `LLM application's function (${name}) type must be required.`,
              );
            if (value.nullable === true)
              output.push(
                `LLM application's function (${name}) type must not be nullable.`,
              );
          }

          // validate function name length and pattern
          const prefix: string = `LLM application's function (${name})`;
          if (/^[0-9]/.test(rawName[0] ?? "") === true)
            output.push(`${prefix} name cannot start with a number.`);
          if (/^[a-zA-Z0-9_-]+$/.test(rawName) === false)
            output.push(
              `${prefix} name must contain only alphanumeric characters, underscores, or hyphens.`,
            );
          if (rawName.length > 64)
            output.push(`${prefix} name cannot exceed 64 characters.`);

          const description: string | undefined = concatDescription(
            JsonApplicationProgrammer.writeDescription({
              description:
                p.description ??
                p.jsDocTags.find((tag) => tag.name === "description")?.text?.[0]
                  ?.text ??
                null,
              jsDocTags: p.jsDocTags,
              kind: "summary",
            }),
          );
          if (description !== undefined && description.length > 1_024)
            output.push(
              `LLM application's function (${name}) description must not exceed 1,024 characters.`,
            );
        }
      }
      if (least === false)
        output.push(
          "LLM application's target type must have at least a function type.",
        );
    }
    return output;
  };

  /**
   * Validates that a {@link MetadataSchema} represents a valid
   * {@link ILlmSchema.IParameters} type: a single, non-nullable, non-optional
   * object type without dynamic property keys.
   */
  const validateObjectSchema = (
    prefix: string,
    label: string,
    schema: MetadataSchema,
  ): string[] => {
    const errors: string[] = [];
    if (schema.isRequired() === false)
      errors.push(
        `${prefix} ${label} cannot be optional (union with undefined).`,
      );
    if (schema.nullable === true)
      errors.push(`${prefix} ${label} cannot be nullable.`);
    if (schema.size() !== 1 || schema.objects.length !== 1)
      errors.push(`${prefix} ${label} must be a single object type.`);
    else if (
      schema.objects[0]!.type.properties.some(
        (p) => p.key.isSoleLiteral() === false,
      )
    )
      errors.push(`${prefix} ${label} cannot have dynamic property keys.`);
    return errors;
  };

  const validateFunction = (name: string, func: MetadataFunction): string[] => {
    const messages: string[] = [];
    const prefix: string = `LLM application's function (${JSON.stringify(name)})`;

    // function name
    if (/^[0-9]/.test(name[0] ?? "") === true)
      messages.push(`${prefix} name cannot start with a number.`);
    if (/^[a-zA-Z0-9_-]+$/.test(name) === false)
      messages.push(
        `${prefix} name must contain only alphanumeric characters, underscores, or hyphens.`,
      );
    if (name.length > 64)
      messages.push(`${prefix} name cannot exceed 64 characters.`);

    // output
    if (func.output.size() !== 0)
      messages.push(
        ...validateObjectSchema(prefix, "return type", func.output),
      );

    // parameters
    if (func.parameters.length !== 0 && func.parameters.length !== 1)
      messages.push(
        `${prefix} must have exactly one parameter or no parameters.`,
      );
    if (func.parameters.length !== 0)
      messages.push(
        ...validateObjectSchema(prefix, "parameter", func.parameters[0]!.type),
      );
    return messages;
  };

  export const writeApplication = (props: {
    context: ITypiaContext;
    modulo: ts.LeftHandSideExpression;
    metadata: MetadataSchema;
    config?: Partial<
      ILlmSchema.IConfig & {
        equals: boolean;
      }
    >;
    name?: string;
  }): ILlmApplication.__IPrimitive => {
    const metadata: MetadataSchema = MetadataSchema.unalias(props.metadata);

    // collect the first parameter's MetadataParameter for each valid function,
    // keyed by property name, for runtime validator generation
    const functionParameters: Record<string, MetadataParameter> =
      Object.fromEntries(
        metadata.objects[0]!.type.properties.filter(
          (p) =>
            p.key.isSoleLiteral() &&
            p.value.size() === 1 &&
            p.value.nullable === false &&
            p.value.isRequired() === true &&
            MetadataSchema.unalias(p.value).functions.length === 1,
        )
          .filter(
            (p) =>
              p.jsDocTags.find(
                (tag) =>
                  tag.name === "hidden" ||
                  tag.name === "ignore" ||
                  tag.name === "internal",
              ) === undefined,
          )
          .map((p) => [
            p.key.getSoleLiteral()!,
            MetadataSchema.unalias(p.value).functions[0]!.parameters[0]!,
          ]),
      );

    // build JSON Schema application, filtering out @human-tagged parameters
    const errorMessages: string[] = [];
    const application: IJsonSchemaApplication<"3.1"> =
      JsonApplicationProgrammer.writeApplication({
        version: "3.1",
        metadata,
        filter: (p) =>
          p.jsDocTags.some((tag) => tag.name === "human") === false,
      });
    // convert each JSON Schema function to an LLM function
    const functions: Array<Omit<ILlmFunction, "parse" | "coerce"> | null> =
      application.functions.map((func) =>
        writeFunction({
          context: props.context,
          modulo: props.modulo,
          className: props.name,
          config: props.config,
          components: application.components,
          function: func,
          errors: errorMessages,
          parameter: functionParameters[func.name] ?? null,
        }),
      );
    return {
      functions: functions.filter((f) => f !== null),
    };
  };

  const writeFunction = (props: {
    context: ITypiaContext;
    modulo: ts.LeftHandSideExpression;
    components: OpenApi.IComponents;
    function: IJsonSchemaApplication.IFunction<OpenApi.IJsonSchema>;
    parameter: MetadataParameter | null;
    errors: string[];
    className?: string;
    config:
      | Partial<
          ILlmSchema.IConfig & {
            equals: boolean;
          }
        >
      | undefined;
  }): Omit<ILlmFunction, "parse" | "coerce"> | null => {
    const config: ILlmSchema.IConfig = LlmSchemaConverter.getConfig(
      props.config,
    );

    // convert parameters and output schemas
    const parameters: ILlmSchema.IParameters | null = writeParameters({
      ...props,
      config,
      accessor: `$input.${props.function.name}.parameters`,
    });
    if (parameters === null) return null;
    const output: ILlmSchema.IParameters | null | undefined = writeOutput({
      config,
      components: props.components,
      schema: props.function.output?.schema ?? null,
      errors: props.errors,
      accessor: `$input.${props.function.name}.output`,
    });
    if (output === null) return null;

    // fall back output description from the function metadata
    if (
      output &&
      output.description === undefined &&
      !!props.function.output?.description?.length
    )
      output.description = props.function.output.description;

    // assemble the LLM function
    return {
      name: props.function.name,
      parameters,
      output: output ?? undefined,
      description: concatDescription({
        summary: props.function.summary,
        description: props.function.description,
      }),
      deprecated: props.function.deprecated,
      tags: props.function.tags,
      validate: writeValidator({
        context: props.context,
        modulo: props.modulo,
        parameter: props.parameter,
        name: props.function.name,
        className: props.className,
        equals: props.config?.equals ?? false,
      }),
    };
  };

  const writeParameters = (props: {
    components: OpenApi.IComponents;
    function: IJsonSchemaApplication.IFunction<OpenApi.IJsonSchema>;
    errors: string[];
    accessor: string;
    config: ILlmSchema.IConfig;
  }): ILlmSchema.IParameters | null => {
    // extract the first parameter's object schema, defaulting to empty object
    const schema: OpenApi.IJsonSchema.IObject | OpenApi.IJsonSchema.IReference =
      (props.function.parameters[0]?.schema as
        | OpenApi.IJsonSchema.IObject
        | OpenApi.IJsonSchema.IReference) ?? {
        type: "object",
        properties: {},
        additionalProperties: false,
        required: [],
      };
    return convertParameters({
      config: props.config,
      components: props.components,
      schema: {
        ...schema,
        title: schema.title ?? props.function.parameters[0]?.title,
        description:
          schema.description ?? props.function.parameters[0]?.description,
      },
      accessor: props.accessor,
      errors: props.errors,
    });
  };

  const writeOutput = (props: {
    components: OpenApi.IComponents;
    config: ILlmSchema.IConfig;
    schema: OpenApi.IJsonSchema | null;
    errors: string[];
    accessor: string;
  }): ILlmSchema.IParameters | null | undefined => {
    if (props.schema === null) return undefined;
    return convertParameters({
      config: props.config,
      components: props.components,
      schema: props.schema as
        | OpenApi.IJsonSchema.IObject
        | OpenApi.IJsonSchema.IReference,
      accessor: props.accessor,
      errors: props.errors,
    });
  };

  /**
   * Converts an OpenAPI object schema to {@link ILlmSchema.IParameters},
   * collecting transformation errors into the provided array.
   */
  const convertParameters = (props: {
    config: ILlmSchema.IConfig;
    components: OpenApi.IComponents;
    schema: OpenApi.IJsonSchema.IObject | OpenApi.IJsonSchema.IReference;
    accessor: string;
    errors: string[];
  }): ILlmSchema.IParameters | null => {
    const result: IResult<ILlmSchema.IParameters, IJsonSchemaTransformError> =
      LlmSchemaConverter.parameters({
        config: props.config,
        components: props.components,
        schema: props.schema,
        accessor: props.accessor,
      });
    if (result.success === false) {
      props.errors.push(
        ...result.error.reasons.map((r) => `  - ${r.accessor}: ${r.message}`),
      );
      return null;
    }
    return result.value;
  };

  const writeValidator = (props: {
    context: ITypiaContext;
    modulo: ts.LeftHandSideExpression;
    parameter: MetadataParameter | null;
    name: string;
    equals: boolean;
    className?: string;
  }): ((props: unknown) => IValidation<unknown>) => {
    // no parameter metadata — generate a permissive `any` validator
    if (props.parameter === null)
      return ValidateProgrammer.write({
        ...props,
        type: props.context.checker.getTypeFromTypeNode(
          TypeFactory.keyword("any"),
        ),
        config: {
          equals: props.equals,
        },
        name: undefined,
      }) as any;

    // unreachable — tsType is always set by the transformer
    if (props.parameter.tsType === undefined)
      throw new Error(
        "Failed to write LLM application's function validator. You don't have to call `LlmApplicationOfValidator.write()` function by yourself, but only by the `typia.llm.applicationOfValidate()` function.",
      );

    // generate a typed validator from the parameter's TypeScript type
    return ValidateProgrammer.write({
      ...props,
      type: props.parameter.tsType!,
      config: {
        equals: props.equals,
      },
      name: props.className
        ? `Parameters<${props.className}[${JSON.stringify(props.name)}]>[0]`
        : undefined,
    }) satisfies ts.CallExpression as any as (
      props: unknown,
    ) => IValidation<unknown>;
  };
}

/**
 * Concatenates summary and description into a single string.
 *
 * If both are present, joins them with a period and double newline, avoiding
 * duplication when the description already starts with the summary.
 */
const concatDescription = (p: {
  summary?: string | undefined;
  description?: string | undefined;
}): string | undefined => {
  if (!p.summary?.length || !p.description?.length)
    return p.summary || p.description;
  const summary: string = p.summary.endsWith(".")
    ? p.summary.slice(0, -1)
    : p.summary;
  return p.description.startsWith(summary)
    ? p.description
    : summary + ".\n\n" + p.description;
};

File: packages/typia/native/core/programmers/llm/LlmApplicationProgrammer.go
package llm

import (
	"fmt"
	"regexp"

	shimast "github.com/microsoft/typescript-go/shim/ast"
	nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
	nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
	nativeprogrammers "github.com/samchon/typia/packages/typia/native/core/programmers"
	nativeiterate "github.com/samchon/typia/packages/typia/native/core/programmers/iterate"
	nativejson "github.com/samchon/typia/packages/typia/native/core/programmers/json"
	schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

type llmApplicationProgrammerNamespace struct{}

var LlmApplicationProgrammer = llmApplicationProgrammerNamespace{}

type LlmApplicationProgrammer_IProps struct {
	Context nativecontext.ITypiaContext
	Modulo  *shimast.Node
	Type    any
	Name    *string
	Init    *shimast.Node
	Config  map[string]any
}

type LlmApplicationProgrammer_IWriteProps struct {
	Context        nativecontext.ITypiaContext
	Modulo         *shimast.Node
	Metadata       *schemametadata.MetadataSchema
	Config         map[string]any
	Name           *string
	ConfigArgument *shimast.Node
}

var llmApplicationProgrammer_factory = shimast.NewNodeFactory(shimast.NodeFactoryHooks{})

func (llmApplicationProgrammerNamespace) Write(props LlmApplicationProgrammer_IWriteProps) *shimast.Node {
	application := LlmApplicationProgrammer.WriteApplication(struct {
		Context  nativecontext.ITypiaContext
		Modulo   *shimast.Node
		Metadata *schemametadata.MetadataSchema
		Config   map[string]any
		Name     *string
	}{
		Context:  props.Context,
		Modulo:   props.Modulo,
		Metadata: props.Metadata,
		Config:   props.Config,
		Name:     props.Name,
	})
	var arguments []*shimast.TypeNode
	if props.Name != nil {
		arguments = []*shimast.TypeNode{llmProgrammer_type_reference(*props.Name)}
	}
	typeNode := llmProgrammer_import_type(props.Context, nativeprogrammers.ImportProgrammer_TypeProps{
		File:      "typia",
		Name:      "ILlmApplication.__IPrimitive",
		Arguments: arguments,
	})
	args := []*shimast.Node{
		llmApplicationProgrammer_factory.NewAsExpression(
			llmApplicationProgrammer_factory.NewSatisfiesExpression(nativefactories.LiteralFactory.Write(application), typeNode),
			typeNode,
		),
	}
	if props.ConfigArgument != nil {
		args = append(args, props.ConfigArgument)
	}
	var typeArguments *shimast.NodeList
	if props.Name != nil {
		typeArguments = llmApplicationProgrammer_factory.NewNodeList([]*shimast.Node{llmProgrammer_type_reference(*props.Name)})
	}
	return llmApplicationProgrammer_factory.NewCallExpression(
		llmProgrammer_internal(props.Context, "llmApplicationFinalize"),
		nil,
		typeArguments,
		llmApplicationProgrammer_factory.NewNodeList(args),
		shimast.NodeFlagsNone,
	)
}

func (llmApplicationProgrammerNamespace) Validate(props struct {
	Config   map[string]any
	Metadata *schemametadata.MetadataSchema
	Explore  nativefactories.MetadataFactory_IExplore
	Top      *schemametadata.MetadataSchema
}) []string {
	if props.Explore.Top == false {
		if props.Top != nil &&
			len(props.Top.Objects) != 0 &&
			props.Explore.Object == props.Top.Objects[0].Type {
			if _, ok := props.Explore.Property.(string); ok &&
				props.Metadata.Size() == 1 &&
				props.Metadata.Nullable == false &&
				props.Metadata.IsRequired() == true &&
				len(props.Metadata.Functions) == 1 {
				return llmApplicationProgrammer_validateFunction(fmt.Sprint(props.Explore.Property), props.Metadata.Functions[0])
			}
		}
		return LlmSchemaProgrammer.Validate(struct {
			Config   map[string]any
			Metadata *schemametadata.MetadataSchema
			Explore  nativefactories.MetadataFactory_IExplore
		}{Config: props.Config, Metadata: props.Metadata, Explore: props.Explore})
	}

	output := []string{}
	isValidType := props.Metadata.Size() == 1 &&
		len(props.Metadata.Objects) == 1 &&
		props.Metadata.IsRequired() == true &&
		props.Metadata.Nullable == false
	if isValidType == false {
		output = append(output, "LLM application's generic argument must be a class/interface type.")
	}
	var object *schemametadata.MetadataObjectType
	if len(props.Metadata.Objects) != 0 {
		object = props.Metadata.Objects[0].Type
	}
	if object != nil {
		for _, property := range object.Properties {
			if property.Key.IsSoleLiteral() == false {
				output = append(output, "LLM application does not allow dynamic keys on class/interface type.")
				break
			}
		}
		least := false
		for _, property := range object.Properties {
			rawName := property.Key.GetSoleLiteral()
			if rawName == nil {
				continue
			}
			value := property.Value
			if len(value.Functions) == 0 {
				continue
			}
			least = true
			name := fmt.Sprintf("%q", *rawName)
			if isValidType == false {
				if len(value.Functions) != 1 || value.Size() != 1 {
					output = append(output, "LLM application's function ("+name+") type does not allow union type.")
				}
				if value.IsRequired() == false {
					output = append(output, "LLM application's function ("+name+") type must be required.")
				}
				if value.Nullable {
					output = append(output, "LLM application's function ("+name+") type must not be nullable.")
				}
			}
			prefix := "LLM application's function (" + name + ")"
			output = append(output, llmApplicationProgrammer_validateName(prefix, *rawName)...)
			description := llmApplicationProgrammer_propertyDescription(property)
			if description != nil && len(*description) > 1024 {
				output = append(output, "LLM application's function ("+name+") description must not exceed 1,024 characters.")
			}
		}
		if least == false {
			output = append(output, "LLM application's target type must have at least a function type.")
		}
	}
	return output
}

func (llmApplicationProgrammerNamespace) WriteApplication(props struct {
	Context  nativecontext.ITypiaContext
	Modulo   *shimast.Node
	Metadata *schemametadata.MetadataSchema
	Config   map[string]any
	Name     *string
}) map[string]any {
	metadata := schemametadata.MetadataSchema_unalias(props.Metadata)
	functionParameters := map[string]*schemametadata.MetadataParameter{}
	if len(metadata.Objects) != 0 {
		for _, property := range metadata.Objects[0].Type.Properties {
			if property.Key.IsSoleLiteral() == false ||
				property.Value.Size() != 1 ||
				property.Value.Nullable ||
				property.Value.IsRequired() == false {
				continue
			}
			unaliased := schemametadata.MetadataSchema_unalias(property.Value)
			if len(unaliased.Functions) != 1 || llmApplicationProgrammer_hidden(property.JsDocTags) {
				continue
			}
			key := property.Key.GetSoleLiteral()
			if key == nil || len(unaliased.Functions[0].Parameters) == 0 {
				continue
			}
			functionParameters[*key] = unaliased.Functions[0].Parameters[0]
		}
	}
	application := nativejson.JsonApplicationProgrammer.WriteApplication(struct {
		Version  string
		Metadata *schemametadata.MetadataSchema
		Filter   func(prop *schemametadata.MetadataProperty) bool
	}{
		Version:  "3.1",
		Metadata: metadata,
		Filter: func(prop *schemametadata.MetadataProperty) bool {
			for _, tag := range prop.JsDocTags {
				if tag.Name == "human" {
					return false
				}
			}
			return true
		},
	})
	components, _ := application["components"].(*nativeiterate.OpenApi_IComponents)
	rawFunctions, _ := application["functions"].([]any)
	functions := []any{}
	for _, raw := range rawFunctions {
		fn, ok := raw.(map[string]any)
		if ok == false {
			continue
		}
		name, _ := fn["name"].(string)
		written := llmApplicationProgrammer_writeFunction(struct {
			Context    nativecontext.ITypiaContext
			Modulo     *shimast.Node
			Components *nativeiterate.OpenApi_IComponents
			Function   map[string]any
			Parameter  *schemametadata.MetadataParameter
			ClassName  *string
			Config     map[string]any
		}{
			Context:    props.Context,
			Modulo:     props.Modulo,
			Components: components,
			Function:   fn,
			Parameter:  functionParameters[name],
			ClassName:  props.Name,
			Config:     props.Config,
		})
		if written != nil {
			functions = append(functions, written)
		}
	}
	return map[string]any{"functions": functions}
}

func llmApplicationProgrammer_writeFunction(props struct {
	Context    nativecontext.ITypiaContext
	Modulo     *shimast.Node
	Components *nativeiterate.OpenApi_IComponents
	Function   map[string]any
	Parameter  *schemametadata.MetadataParameter
	ClassName  *string
	Config     map[string]any
}) map[string]any {
	name, _ := props.Function["name"].(string)
	parameters := llmApplicationProgrammer_writeParameters(struct {
		Context    nativecontext.ITypiaContext
		Components *nativeiterate.OpenApi_IComponents
		Function   map[string]any
		Config     map[string]any
	}{
		Context:    props.Context,
		Components: props.Components,
		Function:   props.Function,
		Config:     props.Config,
	})
	var output any
	if raw, ok := props.Function["output"].(map[string]any); ok {
		if schema, ok := raw["schema"].(nativeiterate.JsonSchema); ok {
			output = llmApplicationProgrammer_writeOutput(struct {
				Context    nativecontext.ITypiaContext
				Components *nativeiterate.OpenApi_IComponents
				Schema     nativeiterate.JsonSchema
				Output     map[string]any
				Config     map[string]any
			}{Context: props.Context, Components: props.Components, Schema: schema, Output: raw, Config: props.Config})
		} else if schema, ok := raw["schema"].(map[string]any); ok {
			output = llmApplicationProgrammer_writeOutput(struct {
				Context    nativecontext.ITypiaContext
				Components *nativeiterate.OpenApi_IComponents
				Schema     nativeiterate.JsonSchema
				Output     map[string]any
				Config     map[string]any
			}{Context: props.Context, Components: props.Components, Schema: nativeiterate.JsonSchema(schema), Output: raw, Config: props.Config})
		}
	}
	result := map[string]any{
		"name":       name,
		"parameters": parameters,
		"validate": llmApplicationProgrammer_writeValidator(struct {
			Context   nativecontext.ITypiaContext
			Modulo    *shimast.Node
			Parameter *schemametadata.MetadataParameter
			Name      string
			Equals    bool
			ClassName *string
		}{
			Context:   props.Context,
			Modulo:    props.Modulo,
			Parameter: props.Parameter,
			Name:      name,
			Equals:    llmApplicationProgrammer_equals(props.Config),
			ClassName: props.ClassName,
		}),
	}
	if output != nil {
		result["output"] = output
	}
	if summary, ok := props.Function["summary"].(string); ok {
		if description, ok := props.Function["description"].(string); ok {
			result["description"] = *llmProgrammer_concat_description(&summary, &description)
		} else {
			result["description"] = summary
		}
	} else if description, ok := props.Function["description"].(string); ok {
		result["description"] = description
	}
	if deprecated, ok := props.Function["deprecated"].(bool); ok && deprecated {
		result["deprecated"] = true
	}
	if tags, ok := props.Function["tags"].([]any); ok && len(tags) != 0 {
		result["tags"] = tags
	}
	return result
}

func llmApplicationProgrammer_writeParameters(props struct {
	Context    nativecontext.ITypiaContext
	Components *nativeiterate.OpenApi_IComponents
	Function   map[string]any
	Config     map[string]any
}) any {
	rawParameters, _ := props.Function["parameters"].([]any)
	if len(rawParameters) == 0 {
		return llmParametersProgrammer_write_parameters_expression(
			props.Context,
			props.Components,
			nativeiterate.JsonSchema{"type": "object", "properties": map[string]any{}, "additionalProperties": false, "required": []any{}},
			props.Config,
		)
	}
	param, _ := rawParameters[0].(map[string]any)
	var schema nativeiterate.JsonSchema
	if raw, ok := param["schema"].(nativeiterate.JsonSchema); ok {
		schema = raw
	} else if raw, ok := param["schema"].(map[string]any); ok {
		schema = nativeiterate.JsonSchema(raw)
	} else {
		schema = nativeiterate.JsonSchema{"type": "object", "properties": map[string]any{}, "additionalProperties": false, "required": []any{}}
	}
	if _, ok := schema["title"]; ok == false {
		if title, ok := param["title"]; ok {
			schema["title"] = title
		}
	}
	if _, ok := schema["description"]; ok == false {
		if description, ok := param["description"]; ok {
			schema["description"] = description
		}
	}
	return llmParametersProgrammer_write_parameters_expression(props.Context, props.Components, schema, props.Config)
}

func llmApplicationProgrammer_writeOutput(props struct {
	Context    nativecontext.ITypiaContext
	Components *nativeiterate.OpenApi_IComponents
	Schema     nativeiterate.JsonSchema
	Output     map[string]any
	Config     map[string]any
}) any {
	if _, exists := props.Schema["description"]; exists == false {
		if desc, ok := props.Output["description"].(string); ok && desc != "" {
			props.Schema["description"] = desc
		}
	}
	return llmParametersProgrammer_write_parameters_expression(props.Context, props.Components, props.Schema, props.Config)
}

func llmApplicationProgrammer_writeValidator(props struct {
	Context   nativecontext.ITypiaContext
	Modulo    *shimast.Node
	Parameter *schemametadata.MetadataParameter
	Name      string
	Equals    bool
	ClassName *string
}) *shimast.Node {
	if props.Parameter == nil || props.Parameter.TsType == nil {
		typ := props.Context.Checker.GetAnyType()
		return nativeprogrammers.ValidateProgrammer.Write(nativeprogrammers.ValidateProgrammer_IProps{
			Context: props.Context,
			Modulo:  props.Modulo,
			Type:    typ,
			Name:    nil,
			Config:  nativeprogrammers.ValidateProgrammer_IConfig{Equals: props.Equals},
		})
	}
	var name *string
	if props.ClassName != nil {
		text := fmt.Sprintf("Parameters<%s[%q]>[0]", *props.ClassName, props.Name)
		name = &text
	}
	return nativeprogrammers.ValidateProgrammer.Write(nativeprogrammers.ValidateProgrammer_IProps{
		Context: props.Context,
		Modulo:  props.Modulo,
		Type:    props.Parameter.TsType,
		Name:    name,
		Config:  nativeprogrammers.ValidateProgrammer_IConfig{Equals: props.Equals},
	})
}

func llmApplicationProgrammer_validateObjectSchema(prefix string, label string, schema *schemametadata.MetadataSchema) []string {
	errors := []string{}
	if schema.IsRequired() == false {
		errors = append(errors, prefix+" "+label+" cannot be optional (union with undefined).")
	}
	if schema.Nullable {
		errors = append(errors, prefix+" "+label+" cannot be nullable.")
	}
	if schema.Size() != 1 || len(schema.Objects) != 1 {
		errors = append(errors, prefix+" "+label+" must be a single object type.")
	} else {
		for _, property := range schema.Objects[0].Type.Properties {
			if property.Key.IsSoleLiteral() == false {
				errors = append(errors, prefix+" "+label+" cannot have dynamic property keys.")
				break
			}
		}
	}
	return errors
}

func llmApplicationProgrammer_validateFunction(name string, fn *schemametadata.MetadataFunction) []string {
	messages := []string{}
	prefix := fmt.Sprintf("LLM application's function (%q)", name)
	messages = append(messages, llmApplicationProgrammer_validateName(prefix, name)...)
	if fn.Output != nil && fn.Output.Size() != 0 {
		messages = append(messages, llmApplicationProgrammer_validateObjectSchema(prefix, "return type", fn.Output)...)
	}
	if len(fn.Parameters) != 0 && len(fn.Parameters) != 1 {
		messages = append(messages, prefix+" must have exactly one parameter or no parameters.")
	}
	if len(fn.Parameters) != 0 {
		messages = append(messages, llmApplicationProgrammer_validateObjectSchema(prefix, "parameter", fn.Parameters[0].Type)...)
	}
	return messages
}

func llmApplicationProgrammer_validateName(prefix string, name string) []string {
	output := []string{}
	if len(name) != 0 && regexp.MustCompile(`^[0-9]`).MatchString(name[:1]) {
		output = append(output, prefix+" name cannot start with a number.")
	}
	if regexp.MustCompile(`^[a-zA-Z0-9_-]+$`).MatchString(name) == false {
		output = append(output, prefix+" name must contain only alphanumeric characters, underscores, or hyphens.")
	}
	if len(name) > 64 {
		output = append(output, prefix+" name cannot exceed 64 characters.")
	}
	return output
}

func llmApplicationProgrammer_propertyDescription(property *schemametadata.MetadataProperty) *string {
	description := property.Description
	if description == nil {
		for _, tag := range property.JsDocTags {
			if tag.Name == "description" && len(tag.Text) != 0 {
				text := tag.Text[0].Text
				description = &text
				break
			}
		}
	}
	desc := nativejson.JsonApplicationProgrammer.WriteDescription(struct {
		Description *string
		JsDocTags   []schemametadata.IJsDocTagInfo
		Kind        string
	}{Description: description, JsDocTags: property.JsDocTags, Kind: "summary"})
	var summary *string
	if s, ok := desc["summary"].(string); ok {
		summary = &s
	}
	if d, ok := desc["description"].(string); ok {
		description = &d
	}
	return llmProgrammer_concat_description(summary, description)
}

func llmApplicationProgrammer_hidden(tags []schemametadata.IJsDocTagInfo) bool {
	for _, tag := range tags {
		if tag.Name == "hidden" || tag.Name == "ignore" || tag.Name == "internal" {
			return true
		}
	}
	return false
}

func llmApplicationProgrammer_equals(config map[string]any) bool {
	if config == nil {
		return false
	}
	value, _ := config["equals"].(bool)
	return value
}
